### PR TITLE
Refactor device source files

### PR DESCRIPTION
**Pull request by vitbaq**
- Change to open each configuration file just once;
- Change data_item to use ell hashmap;
- Separate functions to create a thing from configuration files to another source file;
- Reorder and rename functions to improve readability.

### CODE REVIEW

File: src/properties.c [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+	int sensor_id;
+	int reg_addr;
+	int bit_offset;
+	knot_schema schema;
+	knot_config config;
+
+	data_item_group = get_data_item_groups(fd);
+
+	for (i = 0; data_item_group[i] != NULL ; i++) {
+		rc = set_sensor_id(thing, fd, data_item_group[i], i,
```
Comment (luigiluz) [inline src/properties.c:441]: This 'i' parameter is not necessary anymore

File: src/properties.c [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+	if (rc < 0)
+		return rc;
+
+	storage_close(device_fd);
+
+	return rc;
+}
+
+int properties_create_device(struct knot_thing *thing,
+			     struct device_settings *conf_files)
```
Comment (luigiluz) [inline src/properties.c:562]: Can you check if this column alignment is correct in your text editor?

File: src/properties.h [hunk 3]
```diff
@@ -20,3 +20,7 @@
 
 int properties_create_device(struct knot_thing *thing,
 			     struct device_settings *conf_files);
+
+int properties_clear_credentials(struct knot_thing *thing, char *filename);
+int properties_store_credentials(struct knot_thing *thing, char *filename,
+				 char *id, char *token);
```
Comment (luigiluz) [inline src/properties.h:26]: Can you check if this column alignment is correct in your text editor?

File: src/device.c [hunk 4]
```diff
@@ -635,27 +554,45 @@ static int set_thing_credentials(char *filename)
...
 {
+	int device_fd;
 	int rc;
 
-	rc = set_thing_name(conf_files->device_path);
+	device_fd = storage_open(filename);
+	if (device_fd < 0)
+		return device_fd;
+
+	rc = set_thing_name(device_fd);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 5]
```diff
@@ -153,7 +152,8 @@ static int set_rabbit_mq_url(char *filename)
 	return 0;
 }
 
-static int set_sensor_id(int fd, char *group_id, int index)
+static int set_sensor_id(int fd, char *group_id, int index,
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 6]
```diff
@@ -751,21 +756,58 @@ static void on_config_timeout(int id)
...
+}
 
-	return 0;
+static int create_data_item_polling(void)
+{
+	int success = 0;
+
+	l_hashmap_foreach(thing.data_items, foreach_data_item_polling,
+			  &success);
+	if (!success)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: done

File: confs/device.conf [hunk 7]
```diff
@@ -19,15 +19,10 @@ ModbusURL = tcp://127.0.0.1:502
 
 ####################### KNoT Data Items Parameters #############################
 
-# ATTENTION: Sensor id max value is LIMITED by the number of data items
-# specified in this file. That is:
-# If you specify 3 data items, sensor id max value will be 2 (It is considered
-# that its first possible value is 0).
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply confs/device.conf:25]: Done

File: src/device.c [hunk 8]
```diff
@@ -68,6 +66,8 @@ struct knot_data_item {
 	struct modbus_source modbus_source;
 };
 
+struct knot_thing thing;
+
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 9]
```diff
@@ -240,14 +240,48 @@ static void foreach_config_add_data_item(const void *key, void *value,
 	config_add_data_item(data_item->sensor_id, data_item->config);
 }
 
+static void on_msg_timeout(struct l_timeout *timeout, void *user_data)
+{
+	sm_input_event(EVT_TIMEOUT, user_data);
+}
+
+static int on_modbus_receive(int id)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 10]
```diff
@@ -460,31 +461,35 @@ static int set_data_items(int fd)
...
-		return -EINVAL;
+	thing.data_items = l_hashmap_new();
 
 	data_item_group = get_data_item_groups(fd);
 
 	for (i = 0; data_item_group[i] != NULL ; i++) {
-		rc = set_sensor_id(fd, data_item_group[i], i);
+		data_item_aux = l_new(struct knot_data_item, 1);
+
+		rc = set_sensor_id(fd, data_item_group[i], i, data_item_aux);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 11]
```diff
@@ -751,21 +756,58 @@ static void on_config_timeout(int id)
...
+	knot_msg_schema schema_aux;
+
+	schema_aux.sensor_id = data_item->sensor_id;
+	schema_aux.values = data_item->schema;
+	l_queue_push_head(schema_queue, l_memdup(&schema_aux,
+						 sizeof(knot_msg_schema)));
+}
+
+static void foreach_publish_all_data(const void *key, void *value,
+				     void *user_data)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 12]
```diff
@@ -751,21 +756,58 @@ static void on_config_timeout(int id)
...
+{
+	struct knot_data_item *data_item = value;
+
+	on_publish_data(&data_item->sensor_id, NULL);
+}
+
+static void foreach_config_add_data_item(const void *key, void *value,
+					 void *user_data)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: Done

File: src/device.c [hunk 13]
```diff
@@ -751,21 +756,58 @@ static void on_config_timeout(int id)
...
 
-	for (i = 0; i < thing.data_item_count; i++) {
-		rc = poll_create(DEFAULT_POLLING_INTERVAL, i,
-				 device_read_data);
-		if (rc < 0) {
-			poll_destroy();
-			return rc;
-		}
+static void foreach_data_item_polling(const void *key, void *value,
+				      void *user_data)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c:278]: It is correct in the editor, it must be a GitHub bug.

File: src/device.h [hunk 14]
```diff
@@ -18,12 +18,28 @@
...
 	char *device_path;
 	char *rabbitmq_path;
 };
 
+void device_set_thing_name(struct knot_thing *thing, const char *name);
+void device_set_thing_user_token(struct knot_thing *thing, char *token);
+void device_set_thing_modbus_slave(struct knot_thing *thing, int id, char *url);
+void device_set_new_data_item(struct knot_thing *thing, int sensor_id,
+			      knot_schema schema, knot_config config,
+			      int reg_addr, int bit_offset);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.h:36]: It is correct in the editor, it must be a GitHub bug.

File: src/device.h [hunk 15]
```diff
@@ -18,12 +18,28 @@
...
+void *device_data_item_lookup(struct knot_thing *thing, int *sensor_id);
+void device_set_thing_rabbitmq_url(struct knot_thing *thing, char *url);
+void device_set_thing_credentials(struct knot_thing *thing, const char *id,
+				  const char *token);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.h:40]: It is correct in the editor, it must be a GitHub bug.

File: src/device.h [hunk 16]
```diff
@@ -18,12 +18,28 @@
...
+void device_set_thing_credentials_path(struct knot_thing *thing,
+				       const char *path);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.h:42]: It is correct in the editor, it must be a GitHub bug.

File: src/device.h [hunk 17]
```diff
@@ -18,12 +18,28 @@
...
+
 struct device_settings {
 	char *credentials_path;
 	char *device_path;
 	char *rabbitmq_path;
 };
 
+void device_set_thing_name(struct knot_thing *thing, const char *name);
+void device_set_thing_user_token(struct knot_thing *thing, char *token);
+void device_set_thing_modbus_slave(struct knot_thing *thing, int id, char *url);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.h]: done

File: src/properties.c [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+
+	device_set_thing_rabbitmq_url(thing, rabbitmq_url_aux);
+
+	storage_close(rabbitmq_fd);
+
+	return 0;
+}
+
+static int set_sensor_id(struct knot_thing *thing, int fd, char *group_id,
+			 int index, int *sensor_id)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/properties.c]: done

File: src/properties.c (hunk 1 above)
Comment (vitbaq) [reply src/properties.c]: done

File: src/properties.c (hunk 2 above)
Comment (vitbaq) [reply src/properties.c:570]: It is correct in the editor, it must be a GitHub bug.

File: src/properties.h (hunk 3 above)
Comment (vitbaq) [reply src/properties.h:26]: It is correct in the editor, it must be a GitHub bug.

File: src/device.c [hunk 19]
```diff
@@ -729,19 +735,18 @@ static void on_modbus_disconnected(void *user_data)
 
 static void on_publish_data(void *data, void *user_data)
 {
-	struct knot_data_item data_item;
+	struct knot_data_item *data_item;
 	int *sensor_id = data;
 	int rc;
 
-	if (*sensor_id >= thing.data_item_count)
+	data_item = l_hashmap_lookup(thing.data_items, sensor_id);
```
Comment (kbcvcbk) [inline src/device.c]: Why is the macro `L_INT_TO_PRT` needed on line 165 and 490, but not here?

File: src/device.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (alla3) [reply src/device.c]: It remains on commit c8c435bff16a8f953415fdd5e3e68eb6827fee0e.

File: src/device.c (hunk 19 above)
Comment (vitbaq) [reply src/device.c]: because sensor_id is a prt already.

File: src/device.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/device.c]: done

**Pull request merged**